### ISSUE TITLE
Possibility of leaking `Ptr C_LZ4Stream` ?

### ISSUE THREAD

**Issue by TheKK** [composewell/streamly-lz4#55]
https://github.com/composewell/streamly-lz4/blob/a929c20fb582da95783f84a48dc174204cb8601d/src/Streamly/Internal/LZ4.hs#L394C56-L394C56

In the code above, `Ptr C_LZ4Stream` would be freed ONLY after the following path (in reversed orderd):
- `c_freeStream` was called
- stream enter state, `CompressDone ctx`
- got `Stream.Stop` from inner stream (The only way to produce `CompressDone ctx`)

Does that mean we'll got memory leak in code like:
```haskell
readFile f & compressChunks cfg speed & take 0 & fold drain :: IO ()
```

since we'll never got `Stream.Stop` from `readFile`, or did I missing other path that would clean up `Ptr C_LZ4Stream` properly?

**Comment by adithyaov** [comment]
`compressChunks` should probably be bracketed. I'm yet to look at the code and reason with it.
If you're interested in working on this (update / testing) do let me know if I can help out with anything.

**Comment by harendra-kumar** [comment]
Yes, it is clearly a leak. `bracketIO` should work fine, there should be minimal perf impact due to bracket as this is a chunked stream. If there is a perf impact, we can also write a lightweight version of bracket which cleans up a resource on GC only since there is no possibility of an exception in this particular case. But that should not be necessary I guess.
